### ISSUE TITLE
[ERROR][BUG][KILL] Bot dies in MD version

### ISSUE THREAD

**Issue by jrCleber** [adiwajshing/Baileys#1220]
After a two-minute time the user purposely manually disconnects from the device, the application dies and it is not possible to catch the error.

```javascript
Error: Timed Out
    at C:\Projects\node\chat-api-bot\node_modules\@adiwajshing\baileys\lib\Utils\generics.js:174:36
    at runNextTicks (node:internal/process/task_queues:61:5)
    at listOnTimeout (node:internal/timers:526:9)
    at processTimers (node:internal/timers:500:7) {
  data: {
    stack: 'Error\n' +
      '    at Object.promiseTimeout (C:\\Projects\\node\\chat-api-bot\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\generics.js:168:19)\n' +     
      '    at waitForMessage (C:\\Projects\\node\\chat-api-bot\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:107:42)\n' +
      '    at query (C:\\Projects\\node\\chat-api-bot\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:130:22)\n' +
      '    at C:\\Projects\\node\\chat-api-bot\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:244:38\n' +
      '    at async Object.transaction (C:\\Projects\\node\\chat-api-bot\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\auth-utils.js:78:21)\n' + 
      '    at async resyncAppState (C:\\Projects\\node\\chat-api-bot\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:210:9)\n' +
      '    at async C:\\Projects\\node\\chat-api-bot\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:585:17\n' +
      '    at async C:\\Projects\\node\\chat-api-bot\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\make-mutex.js:11:21\n' +
      '    at async C:\\Projects\\node\\chat-api-bot\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\make-mutex.js:11:21\n' +
      '    at async C:\\Projects\\node\\chat-api-bot\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\make-mutex.js:11:21'
  },
  isBoom: true,
  isServer: false,
  output: {
    statusCode: 408,
    payload: {
      statusCode: 408,
      error: 'Request Time-out',
      message: 'Timed Out'
    },
    headers: {}
  }
}
error Command failed with exit code 1.
```

**Comment by MAYKELL07** [comment]
i get the same error, i wonder why?

**Comment by ookamiiixd** [comment]
Faced the same issue, is there any solution?

**Comment by Bloxoid123** [comment]
the same here, any solution?

**Comment by alsak0de** [comment]
+1

**Comment by canove** [comment]
Same here.

To reproduce:

1 - Read QR code and get connected.
2 - On device, go to "Connected devices"  and disconnect Baileys.
3 - Wait a few minutes. Timed out error will show up.

If you are using Node 15+, unhandled promise rejections kills the process.

`[DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.`

The "future" is now :)

**Comment by HUGEIT** [comment]
can you post a complete trace? if there is any logout event prior to that we could simply cancel the mutex from there but I would need more information

**Comment by banyu-wigara** [comment]
> can you post a complete trace? if there is any logout event prior to that we could simply cancel the mutex from there but I would need more information

Here is the log,
Happened at me yesterday,
Anyway, from my observation, it seems like when you logout from the device, wa send you "server_sync", and 401 status after that, and because of that the mutex still processing even when the connection is closed, can't think any solution for now ... But hope this log will help.

`{"level":20,"time":1645795448616,"pid":11056,"hostname":"DESKTOP-94H8L2K","recv":{"from":"@s.whatsapp.net","type":"server_sync","id":"2990807932","t":"1645795447"},"sent":{"id":"2990807932","to":"@s.whatsapp.net","class":"notification","type":"server_sync"},"msg":"sent \"notification\" ack"}
{"level":20,"time":1645795448617,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"entering transaction"}
{"level":30,"time":1645795448618,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"resyncing regular from v64"}
{"level":20,"time":1645795448620,"pid":11056,"hostname":"DESKTOP-94H8L2K","recv":{"from":"@s.whatsapp.net","type":"server_sync","id":"3281251827","t":"1645795447"},"sent":{"id":"3281251827","to":"@s.whatsapp.net","class":"notification","type":"server_sync"},"msg":"sent \"notification\" ack"}
{"level":20,"time":1645795448622,"pid":11056,"hostname":"DESKTOP-94H8L2K","recv":{"from":"@s.whatsapp.net","type":"server_sync","id":"991842062","t":"1645795447"},"sent":{"id":"991842062","to":"@s.whatsapp.net","class":"notification","type":"server_sync"},"msg":"sent \"notification\" ack"}
{"level":20,"time":1645795448623,"pid":11056,"hostname":"DESKTOP-94H8L2K","recv":{"from":"@s.whatsapp.net","type":"server_sync","id":"107544597","t":"1645795447"},"sent":{"id":"107544597","to":"@s.whatsapp.net","class":"notification","type":"server_sync"},"msg":"sent \"notification\" ack"}
{"level":20,"time":1645795448625,"pid":11056,"hostname":"DESKTOP-94H8L2K","recv":{"from":"@s.whatsapp.net","type":"server_sync","id":"1916261898","t":"1645795447"},"sent":{"id":"1916261898","to":"@s.whatsapp.net","class":"notification","type":"server_sync"},"msg":"sent \"notification\" ack"}
{"level":30,"time":1645795448953,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"synced regular to v65"}
{"level":20,"time":1645795448954,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"committing transaction"}
{"level":40,"time":1645795448959,"pid":11056,"hostname":"DESKTOP-94H8L2K","action":null,"id":"regular","msg":"unprocessable update"}
{"level":40,"time":1645795448960,"pid":11056,"hostname":"DESKTOP-94H8L2K","action":{"timestamp":"1645795446399","keyExpiration":{"expiredKeyEpoch":44168}},"id":"regular","msg":"unprocessable update"}
{"level":40,"time":1645795448961,"pid":11056,"hostname":"DESKTOP-94H8L2K","action":null,"msg":"unprocessable update"}
{"level":40,"time":1645795448961,"pid":11056,"hostname":"DESKTOP-94H8L2K","action":{"timestamp":"1645348406296","primaryFeature":{"flags":["reactions_send"]}},"msg":"unprocessable update"}{"level":20,"time":1645795448962,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"entering transaction"}
{"level":30,"time":1645795448963,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"resyncing regular_high from v78"}
{"level":50,"time":1645795449137,"pid":11056,"hostname":"DESKTOP-94H8L2K","error":{"tag":"stream:error","attrs":{"code":"401"},"content":[{"tag":"conflict","attrs":{"type":"device_removed"},"content":null}]},"msg":"stream errored out"}
{"level":30,"time":1645795449139,"pid":11056,"hostname":"DESKTOP-94H8L2K","error":{"data":{"tag":"stream:error","attrs":{"code":"401"},"content":[{"tag":"conflict","attrs":{"type":"device_removed"},"content":null}]},"isBoom":true,"isServer":false,"output":{"statusCode":401,"payload":{"statusCode":401,"error":"Unauthorized","message":"Stream Errored"},"headers":{}}},"msg":"connection closed"}
  app:serve Conn dfae48d0-b7af-47a4-8cbf-731196f49947 => Connection is close +41s        
  app:serve {
  app:serve   error: Error: Stream Errored
  app:serve       at WebSocket.<anonymous> (THIS_IS_MY_DISK_AND_FOLDER\node_modules\@adiwajshing\baileys\lib\Socket\socket.js:437:13)
  app:serve       at WebSocket.emit (node:events:520:28)
  app:serve       at THIS_IS_MY_DISK_AND_FOLDER\node_modules\@adiwajshing\baileys\lib\Socket\socket.js:229:35
  app:serve       at Object.decodeFrame (THIS_IS_MY_DISK_AND_FOLDER\node_modules\@adiwajshing\baileys\lib\Utils\noise-handler.js:143:17)
  app:serve       at WebSocket.onMessageRecieved (THIS_IS_MY_DISK_AND_FOLDER\node_modules\@adiwajshing\baileys\lib\Socket\socket.js:207:15)
  app:serve       at WebSocket.emit (node:events:520:28)
  app:serve       at Receiver.receiverOnMessage (THIS_IS_MY_DISK_AND_FOLDER\node_modules\ws\lib\websocket.js:1137:20)
  app:serve       at Receiver.emit (node:events:520:28)
  app:serve       at Receiver.dataMessage (THIS_IS_MY_DISK_AND_FOLDER\node_modules\ws\lib\receiver.js:513:14)
  app:serve       at Receiver.getData (THIS_IS_MY_DISK_AND_FOLDER\node_modules\ws\lib\receiver.js:446:17) {
  app:serve     data: { tag: 'stream:error', attrs: [Object], content: [Array] },        
  app:serve     isBoom: true,
  app:serve     isServer: false,
  app:serve     output: { statusCode: 401, payload: [Object], headers: {} }
  app:serve   },
  app:serve   date: 2022-02-25T13:24:09.141Z
  app:serve } +1ms
  app:serve Conn dfae48d0-b7af-47a4-8cbf-731196f49947 => Closed! +14ms
{"level":30,"time":1645795449157,"pid":11056,"hostname":"DESKTOP-94H8L2K","error":{"data":null,"isBoom":true,"isServer":false,"output":{"statusCode":428,"payload":{"statusCode":428,"error":"Precondition Required","message":"Connection Terminated by Server"},"headers":{}}},"msg":"connection closed"}
{"level":50,"time":1645795508967,"pid":11056,"hostname":"DESKTOP-94H8L2K","trace":"Error: Timed Out\n    at THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\generics.js:178:32","node":{"tag":"notification","attrs":{"from":"@s.whatsapp.net","type":"server_sync","id":"3281251827","t":"1645795447"},"content":[{"tag":"collection","attrs":{"name":"regular_high","version":"79"},"content":null}]},"msg":"failed to sync state"}
{"level":20,"time":1645795508969,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"entering transaction"}
{"level":30,"time":1645795508970,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"resyncing regular_low from v65"}
{"level":50,"time":1645795508973,"pid":11056,"hostname":"DESKTOP-94H8L2K","trace":"Error: Connection Closed\n    at sendRawMessage (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:63:19)\n    at sendNode (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:71:16)\n    at query (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:131:15)\n    at THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:241:38\n    at async Object.transaction (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\auth-utils.js:78:21)\n    at async resyncAppState (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:207:9)\n    at async THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:582:17\n    at async THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\make-mutex.js:12:21\n    at async THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\make-mutex.js:12:21","node":{"tag":"notification","attrs":{"from":"@s.whatsapp.net","type":"server_sync","id":"991842062","t":"1645795447"},"content":[{"tag":"collection","attrs":{"name":"regular_low","version":"66"},"content":null}]},"msg":"failed to sync state"}
{"level":20,"time":1645795508977,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"entering transaction"}
{"level":30,"time":1645795508978,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"resyncing critical_block from v64"}
{"level":50,"time":1645795508981,"pid":11056,"hostname":"DESKTOP-94H8L2K","trace":"Error: Connection Closed\n    at sendRawMessage (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:63:19)\n    at sendNode (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:71:16)\n    at query (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:131:15)\n    at THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:241:38\n    at async Object.transaction (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\auth-utils.js:78:21)\n    at async resyncAppState (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:207:9)\n    at async THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:582:17\n    at async THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\make-mutex.js:12:21","node":{"tag":"notification","attrs":{"from":"@s.whatsapp.net","type":"server_sync","id":"107544597","t":"1645795447"},"content":[{"tag":"collection","attrs":{"name":"critical_block","version":"65"},"content":null}]},"msg":"failed to sync state"}
{"level":20,"time":1645795508984,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"entering transaction"}
{"level":30,"time":1645795508985,"pid":11056,"hostname":"DESKTOP-94H8L2K","msg":"resyncing critical_unblock_low from v63"}
{"level":50,"time":1645795508987,"pid":11056,"hostname":"DESKTOP-94H8L2K","trace":"Error: Connection Closed\n    at sendRawMessage (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:63:19)\n    at sendNode (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:71:16)\n    at query (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:131:15)\n    at THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:241:38\n    at async Object.transaction (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\auth-utils.js:78:21)\n    at async resyncAppState (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:207:9)\n    at async THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:582:17","node":{"tag":"notification","attrs":{"from":"@s.whatsapp.net","type":"server_sync","id":"1916261898","t":"1645795447"},"content":[{"tag":"collection","attrs":{"name":"critical_unblock_low","version":"64"},"content":null}]},"msg":"failed to sync state"}
THIS_IS_MY_DISK_AND_FOLDER\node_modules\@adiwajshing\baileys\lib\Utils\generics.js:178
            .then(() => reject(new boom_1.Boom('Timed Out', {
                               ^

Error: Timed Out
    at THIS_IS_MY_DISK_AND_FOLDER\node_modules\@adiwajshing\baileys\lib\Utils\generics.js:178:32 {
  data: {
    stack: 'Error\n' +
      '    at Object.promiseTimeout (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\generics.js:173:19)\n' +
      '    at waitForMessage (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:107:42)\n' +
      '    at query (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\socket.js:130:22)\n' +
      '    at THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:241:38\n' +
      '    at async Object.transaction (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\auth-utils.js:78:21)\n' +
      '    at async resyncAppState (THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:207:9)\n' +
      '    at async THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Socket\\chats.js:582:17\n' +
      '    at async THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\make-mutex.js:12:21\n' +
      '    at async THIS_IS_MY_DISK_AND_FOLDER\\node_modules\\@adiwajshing\\baileys\\lib\\Utils\\make-mutex.js:12:21'
  },
  isBoom: true,
  isServer: false,
  output: {
    statusCode: 408,
    payload: {
      statusCode: 408,
      error: 'Request Time-out',
      message: 'Timed Out'
    },
    headers: {}
  }
}
[nodemon] app crashed - waiting for file changes before starting...`

**Comment by Bloxoid123** [comment]
a bit "dirty" solution, I downgraded to node 14.x.x and the error no longer stops the application.
He only mentions it.
 however I think that by modifying the code as mentioned here, it can be solved in node 15+

https://stackoverflow.com/questions/53940043/unhandledpromiserejectionwarning-this-error-originated-either-by-throwing-insid

**Comment by HUGEIT** [comment]
@banyu-wigara thank you for that.

we have two options, clear the mutex on `device_removed` or handle exceptions around `waitForMessage` so we will call `end()` instead which would make the connection re-launch while having the process running instead of killing the process.

I will think about what would be the better way and once I finished implementing I will suggest these to adi who will eventually decide

**Comment by saifulcoder** [comment]
+1

**Comment by canove** [comment]
remove-stale

**Comment by thuansilva** [comment]
+1

**Comment by edgaru** [comment]
+1

**Comment by HUGEIT** [comment]
I had the time to go into the code just now. maybe there was a much easier solution all along so please try this commit: https://github.com/HUGEIT/Baileys/commit/d2306e5ec7969962d4a711af2171ecfd9464611e

you will still have to handle the logout yourself (dispose sock, create new sock, and so on) but this way, the promises that throw after 1-3 minutes do not crash the process. the old sock simply stalls and the exception gets caught softly.

there still are a bit of annoying logs but making those go away would require some more re-structuring. because I don't want to over-engineer this edge-case I think this is the best option.


I have not made a pull request yet because I'm not sure if the catch can have unwanted implications in other situations but if you test it for a while and throw me an ok sign if everything runs well I will make a pull request to have the change merged into master.

@jrCleber @edgaru @thuansilva etc. please try that.

**Comment by HUGEIT** [comment]
two emojis so far, I interpret that as an ok sign

if you don't report any issues by Wednesday I will PR

**Comment by jrCleber** [comment]
> two emojis so far, I interpret that as an ok sign
> 
> if you don't report any issues by Wednesday I will PR

I started testing today, and it seems to me that it fixed it.  But it's not with all connections that this happens.  And that's what worries me.

**Comment by HUGEIT** [comment]
I think the issue only happens when the phone logs out when an app sync is in progress - for example seconds after scanning 

this is why the error does not come up on all sessions, it's only during these 30 seconds every day when this number has a sync in Baileys going on

**Comment by edgaru** [comment]
> I had the time to go into the code just now. maybe there was a much easier solution all along so please try this commit: [HUGEIT@d2306e5](https://github.com/HUGEIT/Baileys/commit/d2306e5ec7969962d4a711af2171ecfd9464611e)
> 
> you will still have to handle the logout yourself (dispose sock, create new sock, and so on) but this way, the promises that throw after 1-3 minutes do not crash the process. the old sock simply stalls and the exception gets caught softly.
> 
> there still are a bit of annoying logs but making those go away would require some more re-structuring. because I don't want to over-engineer this edge-case I think this is the best option.
> 
> I have not made a pull request yet because I'm not sure if the catch can have unwanted implications in other situations but if you test it for a while and throw me an ok sign if everything runs well I will make a pull request to have the change merged into master.
> 
> @jrCleber @edgaru @thuansilva etc. please try that.

I have two days testing with two instances and it has been working correctly, has not been disconnected.

**Comment by HUGEIT** [comment]
nice thank you @edgaru that sounds great! I have made a PR now, let's see what adi says

**Comment by HUGEIT** [comment]
guys please try the latest commit on the master branch. adi has restructured quite a bit and made some errors not break the process

**Comment by MuhShalah007** [comment]
anyone solve this problem?

**Comment by jefripunza** [comment]
+1 please help me
